### PR TITLE
Leaderelection

### LINKED ISSUE #56: Machine Controller Locking
_From @jessicaochen on January 3, 2018 18:59_

Have leader election/locking so that only one machine controller is running in the cluster at a time.
https://github.com/kubernetes/kube-deploy/blob/master/cluster-api-gcp/machine-controller/controller/machinecontroller.go#L94

For locking see:
- https://github.com/kubernetes/kubernetes/issues/44857
- https://github.com/kubernetes/client-go/tree/master/tools/leaderelection

_Copied from original issue: kubernetes/kube-deploy#469_

### PR DESCRIPTION
**Pull request by medinatiger**
**What this PR does / why we need it**:
This PR added the leader election support in machine controller. We will need this feature in HA configuration for controllers.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #56 

**Special notes for your reviewer**:
I manually tested this feature in single instance case for both leader-elect turned on and off. I deferred the further integration testing for multi-instances cases when we are adding HA configuration in tooling.

**Release note**:

```release-note

```

@kubernetes/kube-deploy-reviewers

### CODE REVIEW

File: cloud/google/cmd/gce-machine-controller/app/controller.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+	leaderElectionClient := kubernetes.NewForConfigOrDie(rest.AddUserAgent(kubeConfig, "machine-leader-election"))
+
+	// Lock required for leader election
+	rl, err := resourcelock.New(
+		leaderElectConfig.ResourceLock,
+		"default",
+		gceMachineControllerName,
+		leaderElectionClient.CoreV1(),
+		resourcelock.ResourceLockConfig{
+			Identity:      id + "-" + gceMachineControllerName,
```
Comment (mvladev) [inline cloud/google/cmd/gce-machine-controller/app/controller.go]: Hostname is not strong unique identifier. Multiple instances can run on the same host (or same named pods in different namespaces). See:

https://github.com/kubernetes/kubernetes/blob/8fcaa19dd06f19d0db803855bc7f1306ba72d953/cmd/kube-controller-manager/app/controllermanager.go#L189-L190

File: cloud/google/cmd/gce-machine-controller/app/controller.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+	if err != nil {
+		return err
+	}
+
+	leaderElectionClient := kubernetes.NewForConfigOrDie(rest.AddUserAgent(kubeConfig, "machine-leader-election"))
+
+	// Lock required for leader election
+	rl, err := resourcelock.New(
+		leaderElectConfig.ResourceLock,
+		"default",
```
Comment (mvladev) [inline cloud/google/cmd/gce-machine-controller/app/controller.go]: Isn't it better to run in the `kube-system` namespace? There are constant for those names in apimachinery - `metav1.NamespaceSystem`

Comment (medinatiger) [reply cloud/google/cmd/gce-machine-controller/app/controller.go]: I have a PR out soon to make namespace configurable for controller to look up/reference resource. Right now everything is assumed to be in default namespace. For this PR, i changed it to metav1.NamespaceSystem as you suggested.

**Comment by krousey** [comment]
I need to dig into this deeper, but a quick question up front: Is there not a way to do this generically for all the controllers? This will also have to be repeated for every provider, and also for pkg/cmd/controller-manager.

**Comment by medinatiger** [comment]
That's price we have to pay when we decided to split gce controller out of controller manager. In theory, we might be able to provide some wrapping in a library. But I didn't see it saves too much code/time. Once it's get LGTMed, I can easily add support for controller-manager. After all, kube scheudler/controller manager has its own copy to deal with leader election with roughly same code. If it's useful, it should be sth in client-go library.

**Comment by medinatiger** [comment]
Also note that all the leader election configuration part will be shared by all controller as a package. We will share this part at least:-)

File: pkg/controller/config/configuration.go [hunk 3]
```diff
@@ -17,22 +17,102 @@ limitations under the License.
 package config
 
 import (
+	"time"
+
 	"github.com/spf13/pflag"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	rl "k8s.io/client-go/tools/leaderelection/resourcelock"
```
Comment (krousey) [inline pkg/controller/config/configuration.go]: There's no real need to alias this import.

File: pkg/controller/config/configuration.go [hunk 4]
```diff
@@ -17,22 +17,102 @@ limitations under the License.
...
 )
 
+// This is a COPY from kubernetes source tree to avoid importing the entire kubernetes tree.
+
+// LeaderElectionConfiguration defines the configuration of leader election
+// clients for components that can run with leader election enabled.
+type LeaderElectionConfiguration struct {
+	// leaderElect enables a leader election client to gain leadership
```
Comment (krousey) [inline pkg/controller/config/configuration.go]: Capitalize the names here to match the change in the field names.

File: cloud/google/cmd/gce-machine-controller/app/controller.go [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+	recorder, err := createRecorder(kubeClientControl)
+	if err != nil {
+		glog.Errorf("Could not create event recorder : %v", err)
+		return err
+	}
+
+	run := func(stop <-chan struct{}) {
+		StartMachineController(server, stop)
+		panic("unreachable")
```
Comment (krousey) [inline cloud/google/cmd/gce-machine-controller/app/controller.go]: Why?

File: cloud/google/cmd/gce-machine-controller/app/controller.go [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+	}
+
+	leaderElectConfig := config.GetLeaderElectionConfig()
+	if !leaderElectConfig.LeaderElect {
+		run(make(<-chan (struct{})))
+		panic("unreachable")
```
Comment (krousey) [inline cloud/google/cmd/gce-machine-controller/app/controller.go]: Again, why?

File: cloud/google/cmd/gce-machine-controller/app/controller.go [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+	// Try and become the leader and start machine controller loops
+	leaderelection.RunOrDie(leaderelection.LeaderElectionConfig{
+		Lock:          rl,
+		LeaseDuration: leaderElectConfig.LeaseDuration.Duration,
+		RenewDeadline: leaderElectConfig.RenewDeadline.Duration,
+		RetryPeriod:   leaderElectConfig.RetryPeriod.Duration,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: run,
+			OnStoppedLeading: func() {
+				glog.Fatalf("leaderelection lost")
```
Comment (krousey) [inline cloud/google/cmd/gce-machine-controller/app/controller.go]: Is it normal to kill yourself when you lose leadership?

File: cloud/google/cmd/gce-machine-controller/app/controller.go [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+			},
+		},
+	})
+	panic("unreachable")
```
Comment (krousey) [inline cloud/google/cmd/gce-machine-controller/app/controller.go]: ?

File: cloud/google/cmd/gce-machine-controller/app/controller.go (hunk 7 above)
Comment (medinatiger) [reply cloud/google/cmd/gce-machine-controller/app/controller.go]: Yes, it's pattern how we use leaderelection package throughout k8s. Here is an example.

https://github.com/kubernetes/kubernetes/blob/7ba97b9200a9cab45f59a81ab8b44fca8ee863eb/cmd/cloud-controller-manager/app/controllermanager.go

File: cloud/google/cmd/gce-machine-controller/app/controller.go (hunk 8 above)
Comment (medinatiger) [reply cloud/google/cmd/gce-machine-controller/app/controller.go]: @krousey , This is pretty much the common pattern how we use those library in controller manager. I simply stick to the example to keep some sort of consistency for coding style.

Comment (medinatiger) [reply cloud/google/cmd/gce-machine-controller/app/controller.go]: Examples:
https://github.com/kubernetes/kubernetes/blob/7ba97b9200a9cab45f59a81ab8b44fca8ee863eb/cmd/cloud-controller-manager/app/controllermanager.go#L194

**Comment by medinatiger** [comment]
@krousey, for your package question, keeping a package to store options for gce machine controller seems to be logical. I don't like the idea of putting everything in main package considering we might add more stuff into controller over time.

File: cloud/google/cmd/gce-machine-controller/app/controller.go (hunk 5 above)
Comment (krousey) [reply cloud/google/cmd/gce-machine-controller/app/controller.go]: Ah I see why now. The `Run` function returns an `error`, and `return nil` would be misleading. However this inner function doesn't have a return value, so the panic isn't needed.

Comment (kris-nova) [reply cloud/google/cmd/gce-machine-controller/app/controller.go]: yeah - why the panic? That seems a bit odd - and the message seems equally as odd

**Comment by krousey** [comment]
I still think it's find putting the options and leader election in package main. However, that is not the pattern that the rest of k8s decided to go (for whatever reason).
/lgtm

**Comment by medinatiger** [comment]
@krousey can you kindly regrant your lgtm label? I have to add a commit to resolve merge conflict from rebasing. But github automatically removed the lgtm label because of it. Thanks!

**Pull request merged**